docs/rfcs/0010-add-state-based-navigation.md

```diff
@@ -0,0 +1,168 @@
+---
+layout: page
+title: Add State-based navigation
+permalink: /rfc/0010-add-state-based-navigation
+---
+
+* Start date: 2023-10-17
+* RFC PR: [4126](https://github.com/mozilla-mobile/firefox-android/pull/4126)
+
+## Summary
+
+Following the acceptance of [0009 - Remove Interactors and Controllers](0009-remove-interactors-and-controllers), Fenix should have a method of navigation that is tied to the `lib-state` model to  provide a method of handling navigation side-effects that is consistent with architectural goals.
+
+## Motivation
+
+Currently, methods of navigation throughout the app are varied. The `SessionControlController` provides [3 examples](https://searchfox.org/mozilla-mobile/rev/aa6bee71a6e0ea73f041a54ddf4d5d4e2f603429/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlController.kt#180) alone:
+
+- `HomeActivity::openToBrowserAndLoad`
+- Calling a `NavController` directly
+- Callbacks like `showTabTray()`
+
+To move to a more consistent Redux-like model, we need a way for features to fire `Action`s and have that result in navigation. This will help decouple our business logic from the Android platform, where a key example of this would be references to the `HomeActivity` throughout the app in order to access the `openToBrowserAndLoad` function.
+
+## Proposal
+
+Moving forward, navigation should be initiated through middlewares that respond to `Action`s. How these middleware handle navigation side-effects can be addressed on a per-case basis, but this proposal includes some generalized advice for 3 common use cases. 
+
+### 1. Screen-based navigation
+
+For screen-based navigation between screens like the settings pages or navigation to the home screen, middlewares should make direct use of a `NavController` that is hosted by the fragment of the current screen's scope.
+
+For a hypothetical example:
+```kotlin
+
+sealed class HistoryAction {
+    object HomeButtonClicked : HistoryAction()
+	data class HistoryGroupClicked(val group: History.Group) : HistoryAction()
+}
+
+class HistoryNavigationMiddleware(private val getNavController: () -> NavController) : Middleware<HistoryState, HistoryAction> {
+	override fun invoke(
+		context: MiddlewareContext<HistoryState, HistoryAction>,
+		next: (HistoryFragmentAction) -> Unit,
+		action: HistoryFragmentAction,
+	) {
+        next(action)
+        when (action) {
+            is HomeButtonClicked -> getNavController().navigate(R.id.home_fragment)
+			is HistoryGroupClicked -> getNavController().navigate(R.id.history_metadata_fragment)
+        }
```

Comment:
Nit: indent missing?

Reply:
Thanks, apparently Android Studio used actual unexpanded tab characters any time I copied a line with a tab in it. Haven't run into a tabs v spaces problem in a longggggg time 😆